forum/backends/mysql/api.py

```diff
@@ -0,0 +1,1115 @@
+        if vote_type not in ["up", "down"]:
+            raise ValueError("Invalid vote_type, use ('up' or 'down')")
+        if not user_vote:
+            user_vote = UserVote.objects.create(user=user, content=content)
+        if vote_type == "up":
+            user_vote.vote = 1
+        else:
+            user_vote.vote = -1
+        user_vote.save()
+        return True
+    else:
+        if user_vote:
+            user_vote.delete()
+            return True
+
+    return False
+
+
+def upvote_content(entity_id: str, entity_type: str, user_id: str) -> bool:
+    """
+    Upvotes the specified thread or comment by the given user.
+
+    Args:
+        thread (dict): The thread or comment data to be upvoted.
+        user (dict): The user who is performing the upvote.
+
+    Returns:
+        bool: True if the vote was successfully updated, False otherwise.
+    """
+    return update_vote(entity_id, entity_type, user_id, vote_type="up")
+
+
+def downvote_content(entity_id: str, entity_type: str, user_id: str) -> bool:
+    """
+    Downvotes the specified thread or comment by the given user.
+
+    Args:
+        thread (dict): The thread or comment data to be downvoted.
+        user (dict): The user who is performing the downvote.
+
+    Returns:
+        bool: True if the vote was successfully updated, False otherwise.
+    """
+    return update_vote(entity_id, entity_type, user_id, vote_type="down")
+
+
+def remove_vote(entity_id: str, entity_type: str, user_id: str) -> bool:
+    """
+    Remove the vote (upvote or downvote) from the specified thread or comment for the given user.
+
+    Args:
+        thread (dict): The thread or comment data from which the vote should be removed.
+        user (dict): The user who is removing their vote.
+
+    Returns:
+        bool: True if the vote was successfully removed, False otherwise.
+    """
+    return update_vote(entity_id, entity_type, user_id, is_deleted=True)
+
+
+def validate_thread_and_user(
+    user_id: str, thread_id: str
+) -> tuple[dict[str, Any], dict[str, Any]]:
+    """
+    Validate thread and user.
+
+    Arguments:
+        user_id (str): The ID of the user making the request.
+        thread_id (str): The ID of the thread.
+
+    Returns:
+        tuple[dict[str, Any], dict[str, Any]]: A tuple containing the user and thread data.
+
+    Raises:
+        ValueError: If the thread or user is not found.
+    """
+    try:
+        thread = CommentThread.objects.get(pk=thread_id)
+        user = ForumUser.objects.get(user__pk=user_id)
+    except ObjectDoesNotExist as exc:
+        raise ValueError("User / Thread doesn't exist") from exc
+
+    return user.to_dict(), thread.to_dict()
+
+
+def pin_unpin_thread(thread_id: str, action: str) -> None:
+    """
+    Pin or unpin the thread based on action parameter.
+
+    Arguments:
+        thread_id (str): The ID of the thread to pin/unpin.
+        action (str): The action to perform ("pin" or "unpin").
+    """
+    try:
```

Comment:
I think it would be better to move this logic to the models and then call the model function from here. What do you think? The same method could be applied on most of the functions like votes, pins, flags, etc.

Reply:
We can indeed move this and other functions directly to models but since we need to keep mongo api and mysql api calls same, it is convenient to have it here. Alternatively, we would have to migrate mongodb api functions to respective models too